ts/WoltLabSuite/Core/Api.ts

```diff
@@ -0,0 +1,294 @@
+import * as Core from "./Core";
+import DomChangeListener from "./Dom/Change/Listener";
+
+type Payload = Record<string, unknown>;
+type ResponseData = {
+  actionName: string;
+  forceBackgroundQueuePerform?: boolean;
+  objectIDs: number[];
+  returnValues: unknown;
+};
+
+export type CallbackFailure = (result: ErrorResult) => boolean;
+
+export enum ErrorCode {
+  CONNECTION_ERROR = "connection_error",
+  EXPECTED_JSON = "expected_json",
+  INVALID_JSON = "invalid_json",
+  STATUS_NOT_OK = "status_not_ok",
+}
+
+export type ErrorResult =
+  | {
+      code: ErrorCode;
+      error?: never;
+      response: Response;
+    }
+  | {
+      code: ErrorCode;
+      error: any;
+      response?: never;
+    };
+
+type ErrorResponsePrevious = {
+  message: string;
+  stacktrace: string;
+};
+
+type ErrorResponse = {
+  exceptionID?: string;
+  file?: string;
+  line?: number;
+  message: string;
+  previous: ErrorResponsePrevious[];
+  returnValues?: {
+    description?: string;
+  };
+  stacktrace?: string;
+};
+
+type RequestBody = {
+  actionName: string;
+  className: string;
+  objectIDs?: number[];
+  parameters?: Payload;
+};
+
+export class Api {
+  private readonly actionName: string;
+  private readonly className: string;
+  private _failure: CallbackFailure | undefined = undefined;
+  private _objectIDs: number[] = [];
+  private _payload: Payload = {};
+  private _showLoadingIndicator = true;
+  private _signal: AbortController | undefined = undefined;
+
+  private constructor(actionName: string, className: string) {
+    this.actionName = actionName;
+    this.className = className;
+  }
+
+  static dboAction(actionName: string, className: string): Api {
+    return new Api(actionName, className);
+  }
+
+  failure(failure: CallbackFailure): this {
+    this._failure = failure;
+
+    return this;
+  }
+
+  getAbortController(): AbortController {
+    if (this._signal === undefined) {
+      this._signal = new AbortController();
+    }
+
+    return this._signal;
+  }
+
+  objectIds(objectIds: number[]): this {
+    this._objectIDs = objectIds;
+
+    return this;
+  }
+
+  payload(payload: Payload): this {
+    this._payload = payload;
+
+    return this;
+  }
+
+  disableLoadingIndicator(): this {
+    this._showLoadingIndicator = false;
+
+    return this;
+  }
+
+  async dispatch(): Promise<unknown> {
+    const url = window.WSC_API_URL + "index.php?ajax-proxy/&t=" + Core.getXsrfToken();
+
+    const body: RequestBody = {
+      actionName: this.actionName,
+      className: this.className,
+    };
+    if (this._objectIDs) {
+      body.objectIDs = this._objectIDs;
+    }
+    if (this._payload) {
+      body.parameters = this._payload;
+    }
+
+    const init: RequestInit = {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
+        "X-Requested-With": "XMLHttpRequest",
+        "X-XSRF-TOKEN": Core.getXsrfToken(),
+      },
+      body: Core.serialize(body),
```

Comment:
[URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) will automatically be correctly serialized by `fetch()`. Maybe it would make sense relying on that and requiring the API user to pass a valid `URLSearchParams` object as the payload?

Reply:
I don't see much value there, at the end of the day it is the same logic as `Core.serialize()`, plus its ergonomics for creating nested structures do not look appealing to me.

I would even go so far to say that the user should not even care what happens, because this is merely an implementation detail. We could transparently change the API to accept a `POST` request with a JSON body instead, but this gets messy when we start requiring non vanilla objects. The current implementation hides the fact _how_ the data gets to the server, it just asks for _what_ should be transmitted.